### PR TITLE
Document Classloaders

### PR DESCRIPTION
**Pull request by tudor-malene**

### CODE REVIEW

File: docs/source/versioning.rst [hunk 1]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
+
+What is different from a traditional application server is that client code - the smart contract logic - is also run when verifying transactions.
+The special requirement for this code is that anytime a transaction is loaded and verified, on any node, it needs to run the exact same code and get the exact same result.
+
+To implement these requirements, there are currently 2 types of classloaders:
+
+
+1. The ``App Classloader``, which is created from all the cordapps deployed on the node.
+This classloader is used when executing the flow logic and when executing rpc commands.
+In a future Corda version it may be split up into isolated Classloaders per Cordapp.
```
Comment (josecoll) [inline docs/source/versioning.rst]: add this as a Note! 
.. note::

File: docs/source/versioning.rst [hunk 2]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
+Anything that is serialized/deserialized and stored on the node is done on this classloader.
```
Comment (josecoll) [inline docs/source/versioning.rst]: done *in* this classloader.

File: docs/source/versioning.rst [hunk 3]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
+
+2. The ``Attachments Classloader``.
+Every time a transaction is sent over the wire to another node, it will be sent together with all its dependencies.
+Which means the receiving node will also download all attachments needed to verify the transaction (and store them in the ``AttachmentsStorage``).
```
Comment (josecoll) [inline docs/source/versioning.rst]: Grammar: cannot start a sentence with "Which".
"This means ..."

File: docs/source/versioning.rst [hunk 4]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
+When the node needs to verify a transaction it will build a new short-lived ``Classloader`` from the attachments listed in the transaction.
+This classloader is used to deserialize the binary format of the transaction and to execute the ``Contract.verify`` code.
+The exact same classloader will be built on any node that verifies that transaction as all attachments are referenced by their cryptographic hash.
```
Comment (josecoll) [inline docs/source/versioning.rst]: Perhaps add a reference to the "No overlap rule" section.

Also worth mentioning what behaviour is executed when multiple inputs with different versions (for a same given contract class) are associated with a transaction.

**Comment by josecoll** [comment]
Please also add an associated JIRA ticket to this PR.

File: docs/source/versioning.rst [hunk 5]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
-.. note:: You can read the original design doc here: :doc:`design/targetversion/design`.
\ No newline at end of file
+.. note:: You can read the original design doc here: :doc:`design/targetversion/design`.
+
+
+
+Classloaders in Corda
+---------------------
+
+Corda is a platform that runs client code, not very different from this point of view from an application server like JBoss.
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply docs/source/versioning.rst]: That's why I avoided to mention `J2EE` or `Websphere`, to not inherit the bad rep .

`Application server` is pretty tame: https://en.wikipedia.org/wiki/Application_server#Application_Server_definition

[the remark this answers is not in the document]
Comment (josecoll) [reply docs/source/versioning.rst]: I wouldn't attempt to make a comparison at all. They are different technologies from different eras.

File: docs/source/versioning.rst [hunk 6]
```diff
@@ -101,4 +101,31 @@ being Corda flows or services.
...
+
+The node administrator "deploys" cordapps in the ``cordapps`` folder. These cordapps register endpoints and logic that the node will execute.
```
[the remark this answers is not in the document]
Comment (tudor-malene) [reply docs/source/versioning.rst]: All that is covered in the flows doc. This just attempts to set the stage for the explanation on how classloaders are used.

**Comment by tudor-malene** [comment]
Superseded by more comprehensive document

**Pull request closed without merging**